File: main/src/ui/java/freedom/app/tcpserver/FileChaCha20.kt
package freedom.app.tcpserver

import org.bouncycastle.jce.provider.BouncyCastleProvider
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.Security
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * ChaCha20-Poly1305 file encryption.
 * A fresh random 32-byte key is generated per file transfer and sent to the
 * receiver over the already-encrypted OTP message channel.
 */
object FileChaCha20 {
    const val KEY_BYTES = 32
    private const val NONCE_BYTES = 12

    init {
        if (Security.getProvider("BC") == null) {
            Security.addProvider(BouncyCastleProvider())
        }
    }

    fun generateKey(): ByteArray =
        ByteArray(KEY_BYTES).also { SecureRandom().nextBytes(it) }

    /**
     * Deterministic 12-byte nonce from fileId + chunkIdx.
     * Safe because the key is unique per file and chunkIdx is unique within a file.
     */
    fun deriveNonce(fileId: String, chunkIdx: Int): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(fileId.toByteArray(Charsets.UTF_8))
        md.update(
            byteArrayOf(
                (chunkIdx shr 24).toByte(),
                (chunkIdx shr 16).toByte(),
                (chunkIdx shr 8).toByte(),
                chunkIdx.toByte()
            )
        )
        return md.digest().copyOf(NONCE_BYTES)
    }

    fun encrypt(plaintext: ByteArray, key: ByteArray, fileId: String, chunkIdx: Int): ByteArray {
        val nonce = deriveNonce(fileId, chunkIdx)
        val cipher = Cipher.getInstance("ChaCha20-Poly1305", "BC")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "ChaCha20"), IvParameterSpec(nonce))
        return cipher.doFinal(plaintext)
    }

    fun decrypt(cipherWithTag: ByteArray, key: ByteArray, fileId: String, chunkIdx: Int): ByteArray? {
        return try {
            val nonce = deriveNonce(fileId, chunkIdx)
            val cipher = Cipher.getInstance("ChaCha20-Poly1305", "BC")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "ChaCha20"), IvParameterSpec(nonce))
            cipher.doFinal(cipherWithTag)
        } catch (e: Exception) {
            null
        }
    }
}

File: ios/Freedom/Core/Crypto/FileChaCha20.swift
// =============================================================================
//  FileChaCha20 -- ChaCha20-Poly1305 file-chunk encryption
// =============================================================================
//
//  A fresh random 32-byte key is generated per file transfer and sent to the
//  receiver inside the FILE_START message, which itself travels over the
//  already-encrypted OTP message channel.
//
//  Nonce derivation is deterministic:
//      nonce = SHA-256(fileId_utf8 || chunkIdx_4byte_bigendian)[0:12]
//
//  Wire format produced by the caller:
//      FILE_START  -->  INFRA:FILE_START:{fileId}:{totalBytes}:{sha256}:{totalChunks}:{hexKey64}:{filename}
//      FCHUNK      -->  FCHUNK:{fileId}:{chunkIdx}/{totalChunks}:{base64_ciphertext_with_tag}
//
// =============================================================================

import Foundation
import CryptoKit

/// ChaCha20-Poly1305 file encryption helpers (stateless, no files on disk).
enum FileChaCha20 {

    /// Key length in bytes (256-bit).
    static let keyBytes = 32

    /// Generate a cryptographically random 32-byte key.
    static func generateKey() -> Data {
        var bytes = Data(count: keyBytes)
        bytes.withUnsafeMutableBytes { ptr in
            _ = SecRandomCopyBytes(kSecRandomDefault, keyBytes, ptr.baseAddress!)
        }
        return bytes
    }

    /// Deterministic 12-byte nonce derived from fileId + chunkIdx.
    ///
    /// `SHA-256(fileId_utf8 || chunkIdx_4byte_bigendian)[0:12]`
    static func deriveNonce(fileId: String, chunkIdx: Int) -> Data {
        var hasher = SHA256()
        hasher.update(data: Data(fileId.utf8))
        var idx = UInt32(chunkIdx).bigEndian
        hasher.update(data: Data(bytes: &idx, count: 4))
        let hash = hasher.finalize()
        return Data(hash.prefix(12))
    }

    /// Encrypt a plaintext chunk. Returns `ciphertext || tag` (16-byte Poly1305 tag appended).
    static func encrypt(plaintext: Data, key: Data, fileId: String, chunkIdx: Int) -> Data? {
        guard key.count == keyBytes else { return nil }
        let nonce = deriveNonce(fileId: fileId, chunkIdx: chunkIdx)
        do {
            let symmetricKey = SymmetricKey(data: key)
            let chaNonce = try ChaChaPoly.Nonce(data: nonce)
            let sealed = try ChaChaPoly.seal(plaintext, using: symmetricKey, nonce: chaNonce)
            return sealed.ciphertext + sealed.tag
        } catch {
            return nil
        }
    }

    /// Decrypt `ciphertext || tag`. Returns the original plaintext or nil on auth failure.
    static func decrypt(cipherWithTag: Data, key: Data, fileId: String, chunkIdx: Int) -> Data? {
        guard key.count == keyBytes, cipherWithTag.count >= 16 else { return nil }
        let nonce = deriveNonce(fileId: fileId, chunkIdx: chunkIdx)
        do {
            let symmetricKey = SymmetricKey(data: key)
            let chaNonce = try ChaChaPoly.Nonce(data: nonce)
            let ciphertext = cipherWithTag.prefix(cipherWithTag.count - 16)
            let tag = cipherWithTag.suffix(16)
            let sealedBox = try ChaChaPoly.SealedBox(nonce: chaNonce, ciphertext: ciphertext, tag: tag)
            return try ChaChaPoly.open(sealedBox, using: symmetricKey)
        } catch {
            return nil
        }
    }
}
